wireapp/wire-android
# fix(rc): perform migration right after the app update [AR-3075]

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-3075" title="AR-3075" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />AR-3075</a>  Perform migration from scala right after the update
  </td></table>
  <br />
 
















----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

After the app update from the scala to AR, we check and execute the migration process only after the user opens the app (`WireActivityViewModel` checks and opens `MigrationScreen` which schedules the `MigrationWorker`). This creates the situation where if the update happened in the background, the user can even be not aware of that, then he/she won't see any notification until he/she opens the app and completes the migration.

### Solutions

Schedule the migration right after the update, by creating a broadcast receiver which listens to the `MY_PACKAGE_REPLACED` actions. The worker is scheduled in the way that it will retry again if it failed because of the network error but will start only if the connection is restored. The app can be opened even during the migration process - view model listens to the current state of the worker, the user is able also to retry manually the migration if it failed because of the network issue.
Migration worker results are simplified because it returns failure only when there's a network error so now every `Failure` means network problem.
Necessary observers (for notifications and persistent web socket) are moved to a dedicated manager and started in `WireApplication` instead of `WireActivityViewModel` so it doesn't live in a lifecycle of Activity but the whole Application.
Creating a channel for web socket is moved to the service right before creating the notification, so there is no "global" channel that's created independently.

### Testing

#### Test Coverage (Optional)


#### How to Test

Update the app from the scala to AR and don't open it, the migration should be performed automatically.

### Attachments (Optional)

https://user-images.githubusercontent.com/30429749/216694259-deea1fdb-0bc0-433f-9aeb-e9f9b3da54fe.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
